encoding/graph/thing/vertex_attribute.rs

```diff
@@ -809,19 +770,92 @@ impl StringAttributeID {
 pub struct StructAttributeID {
     bytes: [u8; Self::LENGTH],
 }
 
 impl StructAttributeID {
-    const LENGTH: usize = AttributeIDLength::LONG_LENGTH;
+    pub(crate) const LENGTH: usize = AttributeIDLength::SHORT_LENGTH;
+    const ENCODING_HASH_LENGTH: usize = { Self::LENGTH - 1 };
+    const ENCODING_STRUCT_TAIL_INDEX: usize = { Self::LENGTH - 1 };
 
     pub fn new(bytes: [u8; Self::LENGTH]) -> Self {
         Self { bytes }
     }
 
-    pub(crate) fn bytes_ref(&self) -> &[u8] {
+    pub fn bytes(&self) -> [u8; Self::LENGTH] {
+        self.bytes
+    }
+
+    pub fn bytes_ref(&self) -> &[u8] {
         &self.bytes
     }
+
+    pub(crate) fn build_hashed_id<const INLINE_LENGTH: usize, Snapshot>(
+        type_id: TypeID,
+        struct_bytes: StructBytes<'_, INLINE_LENGTH>,
+        snapshot: &Snapshot,
+        hasher: &impl Fn(&[u8]) -> u64,
+    ) -> Result<Self, Arc<SnapshotIteratorError>>
+    where
+        Snapshot: ReadableSnapshot,
+    {
+        let key_without_hash = AttributeVertex::build_prefix_type(Prefix::VertexAttributeStruct, type_id);
+        let existing_or_new = Self::find_existing_or_next_disambiguated_hash(
+            snapshot,
+            hasher,
+            key_without_hash.into_bytes(),
+            struct_bytes.bytes().bytes(),
+        )?;
+        match existing_or_new {
+            Either::First(hashed_id) | Either::Second(hashed_id) => Ok(Self { bytes: hashed_id }),
+        }
+    }
+
+    pub(crate) fn find_hashed_id<const INLINE_LENGTH: usize, Snapshot>(
+        type_id: TypeID,
+        struct_bytes: StructBytes<'_, INLINE_LENGTH>,
+        snapshot: &Snapshot,
+        hasher: &impl Fn(&[u8]) -> u64,
+    ) -> Result<Option<Self>, Arc<SnapshotIteratorError>>
+    where
+        Snapshot: ReadableSnapshot,
+    {
+        let key_without_hash = AttributeVertex::build_prefix_type(Prefix::VertexAttributeStruct, type_id);
+        let existing_or_new = Self::find_existing_or_next_disambiguated_hash(
+            snapshot,
+            hasher,
+            key_without_hash.into_bytes(),
+            struct_bytes.bytes().bytes(),
+        )?;
+        match existing_or_new {
+            Either::First(hashed_id) => {
+                debug_assert!(
+                    hashed_id[Self::HASHID_DISAMBIGUATOR_BYTE_INDEX] & Self::HASHID_DISAMBIGUATOR_BYTE_IS_HASH_FLAG
+                        != 0
+                );
+                Ok(Some(Self { bytes: hashed_id }))
+            }
+            Either::Second(hashed_id) => Ok(None),
+        }
+    }
+
+    pub fn get_hash_hash(&self) -> [u8; Self::ENCODING_HASH_LENGTH] {
+        self.bytes[0..Self::ENCODING_HASH_LENGTH].try_into().unwrap()
+    }
+
+    pub fn get_hash_disambiguator(&self) -> u8 {
+        self.bytes[Self::ENCODING_STRUCT_TAIL_INDEX] & !Self::HASHID_DISAMBIGUATOR_BYTE_IS_HASH_FLAG
+    }
+}
+
+impl DisambiguatingHashedID<{ StructAttributeID::ENCODING_HASH_LENGTH + 1 }> for StructAttributeID {
```

Comment:
To be honest I think we can always just drop the `Disambiguated` part of the names, since we only have hash _and_ disambiguated combined, and I think Hashed is shorter + also telling